Add --subnets and --utility-subnets to kops create cluster

meringu opened the pull request:
This change adds two new options to `kops create cluster`

When specifying `--vpc`, `--subnets` can be specified as an unordered array of subnet ids. Kops will then look up the zones of the subnets to find which zone to add the subnet id to.

If `--topology private` is also specified, `--utility-subnets` can similarly be specified.

~If a zone was specified but a subnet wasn't given that matches the zone, then the subnet will be allocated a CIDR with the current behaviour.~ This case fails validation here https://github.com/kubernetes/kops/blob/7bd0a6a703e4f38c62f9c69bcf8a73f1daaa751e/pkg/apis/kops/validation/validation.go#L151

I can add unit tests and docs changes if required, but I am keen to get feedback before I proceed much further.

I have only added support for AWS.

I have tested this by running a command similar to this:

```bash
kops create cluster \
  --zones=us-east-1a,us-east-1b,us-east-1c \
  --topology private \
  --master-zones=us-east-1a,us-east-1b,us-east-1c  \
  --vpc $vpc_id \
  --subnets subnet-111111,subnet-222222,subnet-333333 \
  --utility-subnets subnet-444444,subnet-555555,subnet-666666 \
  $cluster_hosted_zone_name
```

And the cluster spec was as expected.

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -497,6 +502,13 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
 	} else {
+		var zoneToSubnetProviderID map[string]string
+		if len(c.Zones) > 0 && len(c.SubnetIDs) > 0 {
+			zoneToSubnetProviderID, err = getZoneToSubnetProviderID(c.VPCID, c.Zones[0][:len(c.Zones[0])-1], c.SubnetIDs)
```
I think we should check that `api.CloudProviderID(cluster.Spec.CloudProvider) == api.CloudProviderAWS`, until we support this flag on other clouds

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -1203,3 +1230,37 @@ func setOverrides(overrides []string, cluster *api.Cluster, instanceGroups []*ap
+func getZoneToSubnetProviderID(VPCID string, region string, subnetIDs []string) (map[string]string, error) {
+	res := make(map[string]string)
+	if VPCID == "" {
+		return res, fmt.Errorf("must specify vpc when specifying subnets")
```
It is super-tempting to try to make VPC optional (because a subnet ID is unique, and implies a VPC) - but we can do that later :-)

justinsb commented:
/ok-to-test

So I actually really like this - setting up shared subnets is something that gives people a lot of trouble.

We do have 2 mechanisms that could be used instead, but I don't think either apply:

* We can use the experimental `--override` flag to allow anything (in theory) to be set in the cluster manifest - but this is still under development primarily because of the difficulty in addressing a specific member of a list, which is precisely what we're trying to do here.
* For more advanced usage, we do encourage building a manifest directly, and creating the objects using `kops create -f` or `kops replace -f`, and we even have templating via `kops toolbox template`.  However this is a big jump for people that want to reuse existing subnets.

For testing I think just create a new test in `cmd/kops/create_cluster_integration_test.go` - which requires copying one of the folders in `tests/integration/create_cluster/` and then verifying that the output cluster & instance group yaml is as you expect.  Should be straightforward (ping me on slack if you need to or if it gets messy!)

Then I'm +1 on getting this in :-)

justinsb commented:
/ok-to-test

meringu commented:
@justinsb, I have had added some integration tests. I got a bit carried away and added ones for:
 - shared VPC
 - shared subnets
 - shared subnets in a private topology
 - shared subnets, and looking up the VPC id

I also added logic for looking up the VPC id from the first subnet the command is given. let me know what you think.

I also changed the docs for running in shared subnets.

This is everything I plan doing for this change. Let me know if there is anything else I can do or there are changes I should make.

meringu commented:
> Should we verify that all subnets specified are in same VPC?

For a create command:

```
kops create cluster --subnets subnet-1,subnet-2 mycluster.example.com
```
where `subnet-1` is a member of `vpc-1`, and `subnet-2` is a member of `vpc-2`, this case will error with:

```
subnet subnet-2 not found in VPC vpc-1,
```

I could add an extra validation to give a more friendly error? say:

```
subnet-1 and subnet-2 must be in the same VPC
```

***

> sophisticated "templating" functions

This sounds cool, do you have a link that I could read up on this?

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -1203,3 +1248,34 @@ func setOverrides(overrides []string, cluster *api.Cluster, instanceGroups []*ap
+
+func getZoneToSubnetProviderID(VPCID string, region string, subnetIDs []string) (map[string]string, error) {
+	res := make(map[string]string)
+	cloudTags := map[string]string{}
```
Nit: (don't bother fixing) it is odd to mix both ways of building a map in subsequent lines - I think they do the same thing :-)

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -1203,3 +1248,34 @@ func setOverrides(overrides []string, cluster *api.Cluster, instanceGroups []*ap
+	cloudTags := map[string]string{}
+	awsCloud, err := awsup.NewAWSCloud(region, cloudTags)
+	if err != nil {
+		return res, fmt.Errorf("error loading cloud: %v", err)
```
Nit: I think it's typically better to `return nil, fmt.Errorf...` because then if you forget to check the error it is usually more obviously.

justinsb commented:
This is great thank you @meringu .  We can iterate a bit on this - we can see how those error messages actually "feel" in practice, and maybe look at moving some of the code into the AWS specific folder, but I'm generally of the opinion that this looks great and we should merge it in and do additional changes incrementally.

Great contribution :-)

/lgtm

The templating approaches that @geojaz mentioned are I think the ones that I mentioned here: https://github.com/kubernetes/kops/pull/4061#issuecomment-351601080

meringu commented:
Thanks @justinsb and @geojaz!
